### PR TITLE
stats: control collection of feedback

### PR DESCRIPTION
**Pull request by lamxTyler**
We do not need to collect feedback on all requests. This PR control the collection of feedback so that the following cases are not collected:

- Double read for index lookup;

- The number of scan range exceeds a limit because it may affect the performance;

- The request does not pass the probabilistic sampler.

PTAL @coocood @winoros @shenli

### CODE REVIEW

File: statistics/feedback.go [hunk 1]
```diff
@@ -58,6 +59,26 @@ func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool)
 	}
 }
 
+var (
+	// MaxNumberOfRanges is the max number of ranges before split to collect detailed range counts.
+	MaxNumberOfRanges = 20
```
Comment (winoros) [inline statistics/feedback.go]: Why we add this limit?

**Comment by shenli** [comment]
Why not collection stats feedback when`The number of scan range exceeds a limit.`?

File: executor/builder.go [hunk 2]
```diff
@@ -1251,6 +1251,8 @@ func buildNoRangeTableReader(b *executorBuilder, v *plan.PhysicalTableReader) (*
 	} else {
 		e.feedback = statistics.NewQueryFeedback(ts.Table.ID, ts.Hist, ts.StatsInfo().Count(), ts.Desc)
 	}
+	collect := e.feedback.CollectDetailed(len(ts.Ranges))
```
Comment (shenli) [inline executor/builder.go]: Add Comment here.

File: statistics/feedback.go [hunk 3]
```diff
@@ -58,6 +59,26 @@ func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool)
...
+	FeedbackProbability = 0.0
+)
+
+// CollectDetailed decides whether to collect the detailed range counts.
+func (q *QueryFeedback) CollectDetailed(numOfRanges int) bool {
+	if q.hist == nil || q.hist.Len() == 0 {
+		q.Invalidate()
+		return false
+	}
+	if numOfRanges > MaxNumberOfRanges || rand.Float64() > FeedbackProbability {
```
Comment (shenli) [inline statistics/feedback.go]: Add comment here.

File: config/config.toml.example [hunk 4]
```diff
@@ -139,6 +139,9 @@ stats-lease = "3s"
 # Run auto analyze worker on this tidb-server.
 run-auto-analyze = true
 
+# Probability to use the query feedback to update stats, 0 or 1 for always false/true.
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply config/config.toml.example]: But it does not matter if it exceeds the range, if it is less than 0, it is same as 0; if it is greater than 1, it is the same as 1.

**Comment by lamxTyler** [comment]
@winoros @shenli For example, when the ranges are built from `in` clause and most of the ranges just scan a small amount of data,  the collection of the feedback may affect the performance. I think the more ranges a query contains, the fewer data each range scans, so I add this limit.

File: statistics/feedback.go [hunk 5]
```diff
@@ -58,6 +59,29 @@ func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool)
...
+var (
+	// MaxNumberOfRanges is the max number of ranges before split to collect feedback.
+	MaxNumberOfRanges = 20
+	// FeedbackProbability is the probability to collect the feedback.
+	FeedbackProbability = 0.0
+)
+
+// CollectFeedback decides whether to collect the feedback. It returns false when:
+// 1: the histogram is nil or has no buckets;
+// 2: the number of scan ranges exceeds the limit;
```
Comment (coocood) [inline statistics/feedback.go]: Add the reason why to limit the number of scan ranges.

**Comment by coocood** [comment]
LGTM

File: config/config.toml.example (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackysp) [reply config/config.toml.example]: I think it is friendly to set a probability to [0, 100].

[the remark this answers is not in the document]
Comment (lamxTyler) [reply config/config.toml.example]: I think for probability, [0,1] is more reasonable.

**Comment by lamxTyler** [comment]
PTAL @winoros

**Comment by shenli** [comment]
@winoros PTAL

File: statistics/feedback.go [hunk 6]
```diff
@@ -58,6 +59,29 @@ func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool)
...
+	// FeedbackProbability is the probability to collect the feedback.
+	FeedbackProbability = 0.0
+)
+
+// CollectFeedback decides whether to collect the feedback. It returns false when:
+// 1: the histogram is nil or has no buckets;
+// 2: the number of scan ranges exceeds the limit because it may affect the performance;
+// 3: it does not pass the probabilistic sampler.
+func (q *QueryFeedback) CollectFeedback(numOfRanges int) bool {
+	if q.hist == nil || q.hist.Len() == 0 {
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply statistics/feedback.go]: Yes.

File: statistics/feedback.go [hunk 7]
```diff
@@ -58,6 +59,29 @@ func NewQueryFeedback(tableID int64, hist *Histogram, expected int64, desc bool)
 	}
 }
 
+var (
+	// MaxNumberOfRanges is the max number of ranges before split to collect feedback.
+	MaxNumberOfRanges = 20
```
[the remark this answers is not in the document]
Comment (lamxTyler) [reply statistics/feedback.go]: I think it is not needed now.

**Pull request merged**